### DISCUSSION TITLE
Syntactic distinction for OverloadedRecordUpdate?

### DISCUSSION THREAD

**Discussion by adamgundry** [ghc-proposals/ghc-proposals#506] [Ideas]
I've been thinking further about the plans for `OverloadedRecordUpdate` as described in accepted proposal #282. This has an incomplete implementation as part of GHC 9.2, not yet finished for reasons I don't propose to go into here. All of the following assumes GHC will eventually provide an adequate implementation of `HasField`/`setField`, though this is itself pending a rethink.

Instead, I want to raise the question: should there be a syntactic distinction between "traditional" and "overloaded" record updates? The accepted proposal says not; when the extension is enabled, any record update in the source file will be interpreted as an overloaded update. However, there are circumstances in which traditional updates are more powerful, and no obvious way to support both traditional and overloaded updates with a single syntax. This means the extension is non-conservative and fork-like.

For example, where a record type has higher-rank fields or is a GADT, traditional updates may be accepted in cases where an overloaded update would fail to type-check. Conversely, where two fields with the same label are in scope, an overloaded update may be accepted where a traditional update would be ambiguous.

The proposal [briefly considers the possibility of making a syntactic distinction](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0282-record-dot-syntax.rst#should-a-new-update-syntax-be-added):
> One suggestion is that record updates remain as normal, but `a { .foo = 1 }` be used to indicate the new forms of updates. While possible, we believe that option leads to a confusing result, with two forms of update both of which fail in different corner cases. Instead, we recommend use of `C{foo}` as a pattern (with `-XNamedFieldPuns`) to extract fields if necessary.

However, the proposal still leads to the situation that both forms of update exist; it merely requires that the choice of which to use is made at the module or package level, rather than being determined at the site of the record update itself. Arguably it would be simpler for readers if `a { foo = 1 }` was always a traditional update and (say) `a { .foo = 1 }` was always an overloaded update. Crucially, this change would make the extension conservative (enabling it would not change the interpretation of existing updates) and non-fork-like (because the full language would include both update syntaxes). It has the downside that users need to understand two update syntaxes; but that will inevitably be the case for the foreseeable future.

If one is happy with a syntactic distinction for overloaded updates, one might also choose not to use the `OverloadedRecordUpdate` language extension at all, and instead use a lens/optics library. After all, optics already provide an alternative syntax for updates that is more powerful and compositional than traditional updates. Admittedly, optics are more complex, although it is also conceivable that a simpler library could support updates with a syntax like `a & #foo .= 1`.

Another possibility would be to extend the traditional update syntax to make it easier to disambiguate field names. We could carve out part of the local modules proposal (#283) and allow a datatype name to be used as if it were a module prefix for field names. For example, this would then be accepted:
```hs
data S = MkS { foo :: Int }
data T = MkT { foo :: Bool }

e s = S.foo s
f t = t { T.foo = True }
```

Thus it seems we have some options:
 * Go ahead with `OverloadedRecordUpdate` as planned, accepting the current design.
 * Change `OverloadedRecordUpdate` to use a new syntax for the semantics it introduces. (What should this syntax be?)
 * Deprecate `OverloadedRecordUpdate` and recommend the use of (extended) traditional updates or optics.

Any opinions?

**Comment by simonpj** [comment]
> Another possibility would be to extend the traditional update syntax to make it easier to disambiguate field names. We could carve out part of the local modules proposal (https://github.com/ghc-proposals/ghc-proposals/pull/283) and allow a datatype name to be used as if it were a module prefix for field names.

I strongly believe we should do this.

There are two reasons for all this overloading stuff:
1. It allows you to have many records using the same field name
   ```
   data T1 = MkT1 { x :: Int }
   data T2 = MkT2 { x :: Bool }
   f :: T1 -> T1
   f r = r { x=3 }
   ```
   Here, by much typeclass jiggery pokery, with `HasField` and `SetField`, we can do the update.

2. It allows you to write *one function* that works over *many different record types*:
   ```
   g :: HasField "f" r Int => r -> INt
   g r = r { x = r.x + 1 }
   ```
   This `g` works over all records that have an `x` field.

If our goal was just (1), then using a qualified name is far, far more direct:
```
   f :: T1 -> T1
   f r = r { T1.x = 3 }
```
Relying on type classes and constraint solving for *this* task is like using a tactical nuclear weapon to anihilate a fly.

Now (2) is a worthy goal!  But it comes with consequences, as Adam says: it operates one field at a time, and is much less powerful for GADTs and existentials.  Solving (1) using qualified names is simple, direct, and keeps all the good things about the ``traditional'' story.  It fills out the traditional design.

----------

I see that as pretty orthogonal to Adam's question about whether to have a separate syntax for overloaded record update. Adam you write *"If one is happy with a syntactic distinction for overloaded updates, one might also choose not to use the OverloadedRecordUpdate language extension at all, and instead use a lens/optics library."*  Can you say more? Do you mean "add no new syntax whatsoever; just use ordinary function calls to optics functions for record update"?  Might it not help to have some syntactic support?  That's what `OverloadedRecordDot` bought us.  Or don't we need syntactic support for updates?

**Reply by adamgundry** [reply]
> Adam you write _"If one is happy with a syntactic distinction for overloaded updates, one might also choose not to use the OverloadedRecordUpdate language extension at all, and instead use a lens/optics library."_ Can you say more? Do you mean "add no new syntax whatsoever; just use ordinary function calls to optics functions for record update"? Might it not help to have some syntactic support? That's what `OverloadedRecordDot` bought us. Or don't we need syntactic support for updates?

Right, my point is that since optics give us a good way to write overloaded updates already, it's less clear that it is worthwhile adding special syntax for them. We might choose to do so anyway (e.g. because we can pick nicer syntactic sugar). But while optics are syntactically more noisy, they are ultimately a much more powerful feature.

Unlike `OverloadedRecordDot`, where precedents for the syntactic sugar were very strongly established by other languages, I don't think there are such strong precedents for record update syntax? The main one I'm aware of is "traditional" Haskell update syntax, and that's precisely what we're considering moving away from...

**Reply by AntC2** [reply]
@simonpj's 2. "one function that works over many different record types" precisely relies on "traditional" Haskell update syntax. And it's type-safe because there's the `Hasfield` constraint.

Without `Hasfield`, traditional record update is a partial function: we don't know the constructor/don't know if it has such a field until runtime.

For higher-rank fields or with GADTs, I think it's reasonable to move away from record update in favour of full record build with explicit constructor. I think of those as in effect 'starting over'. I'd say the same for type-changing update.

**Comment by Ericson2314** [comment]
> We could carve out part of the local modules proposal (https://github.com/ghc-proposals/ghc-proposals/pull/283) and allow a datatype name to be used as if it were a module prefix for field names.

I am reminded of https://github.com/ghc-proposals/ghc-proposals/pull/310 where I basically wanted the same thing. Isn't the problem still data families? We have no data type name to qualify the field name with.

**Reply by adamgundry** [reply]
Yes, data families are a problem, in that even if we use the family name as a qualifier, we still get stuck if somebody writes
```hs
data family T a
data instance T Int  = MkTInt  { foo :: Int  }
data instance T Bool = MkTBool { foo :: Bool }
```
...but I'm inclined to just say "don't do that". Users can always write explicit case statements to distinguish them, if they really have to use a single family with duplicated names.

I agree that this idea is similar in spirit to #310, in particular "Namespacing, not overloading". The nice thing about using module-like qualifiers is that we don't need any new syntax, and it works for both selection and update.

**Reply by Ericson2314** [reply]
I suppose in a multiple field update if one just made one field qualified it could figure it out?

**Reply by Ericson2314** [reply]
Anyways, I think it would be great if you were to proposal a new proposal (or just take over #310) and make it do the qualified fields things. I am even OK if that is the only way to get the old semantics, and without qualified fields (and the extension on) it is the new semantics.

This allows punting on accessors (is `T.foo` a thing in general?) And gives us full expressive power with "Namespacing, not overloading" (which I hope can go in principles section as at least an optional principle). I think this is a good first step we can build upon later.

**Reply by Ericson2314** [reply]
Having read https://gitlab.haskell.org/ghc/ghc/-/issues/19463 and https://gitlab.haskell.org/ghc/ghc/-/issues/19461, I like where the plans seem to be going!

- Full separation in surface language and compiler implementation of `HasField` vs flexible but rigorously renamer-resolved field resolution.
- Per https://github.com/ghc-proposals/ghc-proposals/pull/366 `AmbiguousFieldOcc` is going to go, but not as part of the cleanup but after it. This means we can avoid "simplified subsumption"-type problems and have a proper deprecation cycle before removing the mis-feature.

**Reply by adamgundry** [reply]
@Ericson2314 Having thought about this some more in discussion with others, I'm increasingly inclined to change name resolution for record updates to take advantage of type signatures in an extremely limited way. That is, do something like #310 but based on a type signature in the right place, rather than new syntax. I think that ends up being quite a lot simpler than even a fragment of the local modules proposal.

The proposal isn't quite ready yet but in case you are interested, the current draft is at https://github.com/adamgundry/ghc-proposals/blob/disambiguating-record-updates-using-type-signatures/proposals/0000-disambiguating-record-updates-using-type-signatures.rst; I'd particularly like to check I'm representing #310 fairly.

**Reply by simonpj** [reply]
> All this suggests that we should have a mechanism for record updates where the programmer explicitly specifies which type is being updated. 

I agree with this, strongly.

I'm pretty agnostic about the precise mechanism -- the draft suggests three alterantives just before "propsed change spec".   But I can see the attraction of the type-sig approach becuase it avoids new syntax.

Please add section numbers!!

**Reply by adamgundry** [reply]
@simonpj I've added section numbers and done a bit more editing.

**Comment by adamgundry** [comment]
I made some further changes to the draft proposal at https://github.com/adamgundry/ghc-proposals/blob/disambiguating-record-updates-using-type-signatures/proposals/0000-disambiguating-record-updates-using-type-signatures.rst after talking it over with @sheaf. It now affects non-overloaded updates only.

For giving `OverloadedRecordUpdate` a way to fall back on non-overloaded updates in some cases, we could use the presence of a module qualifier. Currently this is rejected, e.g. if `OverloadedRecordUpdate` is enabled then `r { M.f = v }` is an error, because it doesn't make sense to have an overloaded update be restricted based on a module qualifier. But we could instead say that it becomes a non-overloaded update. I think this belongs in a separate proposal though.
